File: src/Modbot.Companion/Presentation/Shortcuts.cs
namespace Modbot.Companion.Presentation;

/// <summary>Where a key is listed on the shortcut sheet.</summary>
public enum ShortcutGroup
{
    General,
    GoTo,
    Lists,
    Filters,
}

/// <summary>
/// One key the window answers to.
/// </summary>
/// <param name="Keys">
/// The keys, lower case, modifiers first: <c>mod+k</c>, <c>?</c>, <c>j</c>, <c>escape</c>,
/// <c>enter</c>, or a chord such as <c>g e</c> (press <c>g</c>, then <c>e</c> within a second).
/// <c>mod</c> is Ctrl; the companion has no Mac build.
/// </param>
/// <param name="Label">What it does, for the sheet and the command palette.</param>
/// <param name="Page">Belongs to the page behind the palette or the sheet: does not fire while the palette or the sheet is open.</param>
/// <param name="Hidden">Left off the sheet and the palette: a second spelling of a key that is already listed.</param>
public sealed record Shortcut(
    string Keys,
    string Label,
    ShortcutGroup Group,
    Action Run,
    bool Page = false,
    bool Hidden = false);

/// <summary>What one key press should do, decided against the registry.</summary>
/// <param name="Run">The shortcut to run, or null.</param>
/// <param name="Pending">The first key of a chord, to hold until the second arrives, or null.</param>
public readonly record struct ShortcutOutcome(Shortcut? Run, string? Pending)
{
    public static ShortcutOutcome Nothing => default;
}

/// <summary>
/// Key presses as tokens, and tokens as words.
/// </summary>
/// <remarks>
/// Shift is not written for printable keys: <c>?</c> is what the key produced, and <c>shift+/</c>
/// is what it took to produce it, which differs between keyboards. Shift is kept for named keys
/// such as <c>shift+enter</c>, where there is no character to speak for it.
/// </remarks>
public static class KeyTokens
{
    /// <summary>
    /// One key press as a token: <c>mod+k</c>, <c>?</c>, <c>j</c>, <c>escape</c>, <c>arrowdown</c>.
    /// </summary>
    /// <param name="key">
    /// The character the key produced (<c>j</c>, <c>?</c>) or the name of a key that produces
    /// none (<c>escape</c>, <c>enter</c>, <c>arrowdown</c>). Null for a modifier on its own.
    /// </param>
    public static string? Token(string? key, bool mod, bool alt, bool shift)
    {
        if (string.IsNullOrEmpty(key))
            return null;

        var printable = key.Length == 1;
        var parts = new List<string>(4);
        if (mod)
            parts.Add("mod");
        if (alt)
            parts.Add("alt");
        if (shift && !printable)
            parts.Add("shift");
        parts.Add(key.ToLowerInvariant());

        return string.Join('+', parts);
    }

    /// <summary>The keys as a person reads them: <c>Ctrl K</c>, <c>G then E</c>, <c>?</c>.</summary>
    public static string Describe(string keys)
    {
        ArgumentNullException.ThrowIfNull(keys);
        return string.Join(" then ", keys.Split(' ', StringSplitOptions.RemoveEmptyEntries).Select(DescribeCombo));
    }

    /// <summary>One combo of a chord: <c>mod+k</c> as <c>Ctrl K</c>.</summary>
    public static string DescribeCombo(string combo)
    {
        ArgumentNullException.ThrowIfNull(combo);
        return string.Join(' ', combo.Split('+').Select(part => part switch
        {
            "mod" => "Ctrl",
            "alt" => "Alt",
            "shift" => "Shift",
            "escape" => "Esc",
            "enter" => "Enter",
            "arrowdown" => "↓",
            "arrowup" => "↑",
            "arrowleft" => "←",
            "arrowright" => "→",
            "space" => "Space",
            _ => part.Length == 1 ? part.ToUpperInvariant() : part,
        }));
    }
}

/// <summary>
/// The keyboard: one registry, one place that decides what a key does.
/// </summary>
/// <remarks>
/// <para>Modelled on the web app's keys (research 2026-09-16): <c>Ctrl+K</c> for the command
/// palette, <c>?</c> for the shortcut sheet, <c>g</c> then a letter to go to a page, <c>j</c>/<c>k</c>
/// to move in a list, <c>Enter</c> to open, <c>Esc</c> to close, <c>f</c> for a filter.</para>
/// <para>Two rules decide everything else here:</para>
/// <list type="bullet">
/// <item><strong>A key typed into a text box is text, never a shortcut.</strong> Only keys with
/// Ctrl held, and Escape, fire while a text box has focus. So <c>j</c> in the pairing box is a
/// letter and <c>Ctrl+K</c> there still opens the palette.</item>
/// <item><strong>A page's keys go quiet while a panel is open.</strong> <c>j</c> and <c>Enter</c>
/// belong to the list behind the palette, and moving that list's selection under a panel
/// nobody can see would be a surprise.</item>
/// </list>
/// <para>Shortcuts are registered in named sets, so a page can replace its own set when it is
/// shown without touching the window's. Later sets shadow earlier ones for the same keys, and
/// within a set the last one added wins.</para>
/// <para>No Avalonia in here, so the rules can be tested without a window.</para>
/// </remarks>
public sealed class ShortcutRegistry
{
    /// <summary>How long the first key of a chord waits for the second.</summary>
    public static readonly TimeSpan ChordTimeout = TimeSpan.FromSeconds(1);

    private readonly List<(string Set, Shortcut Shortcut)> _shortcuts = [];

    private readonly List<string> _sets = [];

    /// <summary>Replaces one named set of shortcuts. An empty list takes the set away.</summary>
    public void Set(string set, IEnumerable<Shortcut> shortcuts)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(set);
        ArgumentNullException.ThrowIfNull(shortcuts);

        _shortcuts.RemoveAll(s => s.Set == set);
        if (!_sets.Contains(set))
            _sets.Add(set);

        foreach (var shortcut in shortcuts)
            _shortcuts.Add((set, shortcut));
    }

    /// <summary>Everything registered, in the order it was registered.</summary>
    public IReadOnlyList<Shortcut> All
        => [.. _sets.SelectMany(set => _shortcuts.Where(s => s.Set == set).Select(s => s.Shortcut))];

    /// <summary>
    /// What the sheet and the palette list: one row per key, the one that would fire, without
    /// the hidden ones, in group order.
    /// </summary>
    public IReadOnlyList<Shortcut> Listed
    {
        get
        {
            var byKeys = new Dictionary<string, Shortcut>(StringComparer.Ordinal);
            foreach (var shortcut in All)
                byKeys[shortcut.Keys] = shortcut;

            return [.. byKeys.Values.Where(s => !s.Hidden).OrderBy(s => s.Group)];
        }
    }

    /// <summary>
    /// Decides what one key press does.
    /// </summary>
    /// <param name="token">The press, from <see cref="KeyTokens.Token"/>.</param>
    /// <param name="pending">The first key of a chord pressed a moment ago, or null.</param>
    /// <param name="typing">Whether a text box has focus.</param>
    /// <param name="panelOpen">Whether the palette or the shortcut sheet is open.</param>
    public ShortcutOutcome Decide(string token, string? pending, bool typing, bool panelOpen)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(token);

        var candidates = All.Where(s => Allowed(s, typing, panelOpen)).ToList();
        var keys = candidates.Select(s => s.Keys).ToList();

        if (pending is not null)
        {
            var chord = $"{pending} {token}";
            if (keys.Contains(chord))
                return new ShortcutOutcome(Last(candidates, chord), null);

            // A chord that went nowhere falls through to the single key, so `g` then `j` still
            // moves the list rather than being swallowed.
        }

        if (keys.Any(k => k.StartsWith($"{token} ", StringComparison.Ordinal)))
            return new ShortcutOutcome(null, token);

        return keys.Contains(token)
            ? new ShortcutOutcome(Last(candidates, token), null)
            : ShortcutOutcome.Nothing;
    }

    private static bool Allowed(Shortcut shortcut, bool typing, bool panelOpen)
    {
        if (typing && !shortcut.Keys.Contains("mod+", StringComparison.Ordinal) && shortcut.Keys != "escape")
            return false;

        return !(panelOpen && shortcut.Page);
    }

    /// <summary>The last one registered wins, so a page can shadow a window key while it is open.</summary>
    private static Shortcut Last(List<Shortcut> candidates, string keys)
        => candidates.Last(s => s.Keys == keys);
}

File: src/Modbot.Web/src/lib/shortcuts.ts
import { useEffect, useRef, useSyncExternalStore } from 'react'

/**
 * The keyboard: one registry, one listener on the window.
 *
 * Modelled on Linear's keys (research 2026-09-16): `Ctrl/Cmd+K` for the command palette, `?` for
 * the shortcut sheet, `g` then a letter to go to a page, `j`/`k` to move in a list, `Enter` to
 * open, `Esc` to close, `f` for the filter bar, `/` for search.
 *
 * Two rules decide everything else here:
 *
 * - **A key typed into a field is text, never a shortcut.** Only keys with Ctrl or Cmd held, and
 *   Escape, fire while an input, a textarea, a select or an editable element has focus. So `j`
 *   in the search box is a letter and `Ctrl+K` in the search box still opens the palette.
 * - **A page's keys go quiet while a popup is open.** `j` and `Enter` belong to the list behind
 *   the popup, and moving that list's selection under a dialog nobody can see would be a surprise.
 *   Dialogs say they are open with `useModal`, and shortcuts marked `page` wait.
 *
 * Every page registers its own keys and unregisters them when it unmounts, so the sheet lists
 * exactly what works on the screen that is open.
 */

export type ShortcutGroup = 'General' | 'Go to' | 'Lists' | 'Filters' | 'Popups'

export type Shortcut = {
  /**
   * The keys, lower case, modifiers first: `mod+k`, `?`, `j`, `escape`, `enter`, or a chord
   * such as `g m` (press `g`, then `m` within a second). `mod` is Ctrl on Windows and Linux and
   * Cmd on a Mac.
   */
  keys: string
  /** What it does, for the sheet and the palette. */
  label: string
  group: ShortcutGroup
  /** Belongs to the page behind any popup: does not fire while one is open. */
  page?: boolean
  /** Left out of the sheet and the palette: a key that only makes sense with the mouse on a row. */
  hidden?: boolean
  run: (event: KeyboardEvent) => void
}

/** How long the first key of a chord waits for the second. */
export const CHORD_TIMEOUT_MS = 1000

// ── The registry ────────────────────────────────────────────────────────────────────────────

let registered: Shortcut[] = []
const listeners = new Set<() => void>()

function notify() {
  for (const l of listeners) l()
}

/** Adds one shortcut; the function returned takes it away again. Later registrations win. */
export function registerShortcut(shortcut: Shortcut): () => void {
  registered = [...registered, shortcut]
  notify()

  return () => {
    registered = registered.filter((s) => s !== shortcut)
    notify()
  }
}

/** Everything registered right now, in registration order. */
export function listShortcuts(): Shortcut[] {
  return registered
}

/** The registered shortcuts as React state, so the sheet redraws when a page changes them. */
export function useShortcutList(): Shortcut[] {
  return useSyncExternalStore(
    (cb) => {
      listeners.add(cb)
      return () => listeners.delete(cb)
    },
    listShortcuts,
    listShortcuts,
  )
}

/**
 * Registers shortcuts for as long as the component is mounted.
 *
 * The list is rebuilt on every render and need not be memoised: what is registered changes only
 * when the keys, labels or groups do, and a key press always runs the handler from the latest
 * render, so a handler may close over state freely.
 */
export function useShortcuts(shortcuts: Shortcut[]): void {
  const latest = useRef(shortcuts)

  useEffect(() => {
    latest.current = shortcuts
  })

  const signature = shortcuts.map((s) => `${s.keys}\t${s.label}\t${s.group}\t${s.page ? 1 : 0}\t${s.hidden ? 1 : 0}`).join('\n')

  useEffect(() => {
    const away = latest.current.map((s, i) =>
      registerShortcut({ ...s, run: (event) => latest.current[i]?.run(event) }),
    )
    return () => away.forEach((off) => off())
  }, [signature])
}

// ── Popups and dialogs ──────────────────────────────────────────────────────────────────────

let modalDepth = 0

/** Says a dialog is open for as long as the component is mounted, so page keys wait. */
export function useModal(): void {
  useEffect(() => {
    modalDepth++
    return () => {
      modalDepth--
    }
  }, [])
}

/**
 * Whether anything modal is open: a popup or dialog that said so, or any Radix dialog or popover
 * on the page, which carry `role="dialog"` while open. The second check catches the confirmation
 * dialogs and pickers that never call `useModal`, so `Enter` on a confirm button is never also
 * `Enter` on the list behind it.
 */
export function isModalOpen(): boolean {
  if (modalDepth > 0) return true
  return typeof document !== 'undefined' && document.querySelector('[role="dialog"][data-state="open"]') !== null
}

// ── Pure parts, kept free of the DOM so they can be tested with Node alone ──────────────────

/** Whether typing into this element is what a key press means. */
export function isTyping(target: {
  tagName?: string
  type?: string
  isContentEditable?: boolean
} | null): boolean {
  if (!target) return false
  if (target.isContentEditable) return true

  const tag = target.tagName?.toUpperCase()
  if (tag === 'TEXTAREA' || tag === 'SELECT') return true
  if (tag !== 'INPUT') return false

  // A checkbox, a radio or a button is an input in name only.
  const type = (target.type ?? 'text').toLowerCase()
  return !['checkbox', 'radio', 'button', 'submit', 'reset', 'range', 'file', 'color'].includes(type)
}

/**
 * One key press as a token: `mod+k`, `?`, `j`, `escape`, `arrowdown`.
 *
 * Shift is not written for printable keys — `?` is what the key produced, and `shift+/` is what
 * it took to produce it, which differs between keyboards. Shift is kept for named keys such as
 * `shift+enter`, where there is no character to speak for it.
 */
export function keyToken(event: {
  key: string
  ctrlKey?: boolean
  metaKey?: boolean
  altKey?: boolean
  shiftKey?: boolean
}): string | null {
  const key = event.key
  if (!key || key === 'Control' || key === 'Meta' || key === 'Alt' || key === 'Shift') return null

  const printable = key.length === 1
  const parts: string[] = []
  if (event.ctrlKey || event.metaKey) parts.push('mod')
  if (event.altKey) parts.push('alt')
  if (event.shiftKey && !printable) parts.push('shift')
  parts.push(key.toLowerCase())

  return parts.join('+')
}

/**
 * Decides what a token does against a list of key strings, given the chord key pressed a
 * moment ago (if any).
 *
 * Returns the matched keys, or the token to hold as the start of a chord, or nothing.
 */
export function matchKeys(
  keys: readonly string[],
  pending: string | null,
  token: string,
): { run?: string; pending?: string } {
  if (pending) {
    const chord = `${pending} ${token}`
    if (keys.includes(chord)) return { run: chord }
    // A chord that went nowhere falls through to the single key, so `g` then `j` still moves
    // the list rather than being swallowed.
  }

  if (keys.some((k) => k.startsWith(`${token} `))) return { pending: token }
  if (keys.includes(token)) return { run: token }

  return {}
}

/**
 * Each key of a chord as a person reads it: `g m` is `['G', 'M']`, `mod+k` is `['Ctrl K']`. Every
 * key hint on screen is spelled from this, so a key never reads `g` in one place and `G` in another:
 * `Kbd` draws one box per name, and the sidebar's go-to chords are the names side by side.
 */
export function keyNames(keys: string, mac: boolean): string[] {
  const names: Record<string, string> = {
    mod: mac ? '⌘' : 'Ctrl',
    alt: mac ? '⌥' : 'Alt',
    shift: 'Shift',
    escape: 'Esc',
    enter: 'Enter',
    arrowdown: '↓',
    arrowup: '↑',
    arrowleft: '←',
    arrowright: '→',
    ' ': 'Space',
  }

  return keys.split(' ').map((combo) =>
    combo
      .split('+')
      .map((part) => names[part] ?? (part.length === 1 ? part.toUpperCase() : part))
      .join(' '),
  )
}

/** The keys as a person reads them: `Ctrl K`, `⌘ K`, `G then M`, `?`. */
export function describeKeys(keys: string, mac: boolean): string {
  return keyNames(keys, mac).join(' then ')
}

// ── The listener ────────────────────────────────────────────────────────────────────────────

let pendingChord: string | null = null
let chordTimer: ReturnType<typeof setTimeout> | null = null

function onKeyDown(event: KeyboardEvent) {
  if (event.defaultPrevented || event.isComposing) return

  const token = keyToken(event)
  if (!token) return

  const typing = isTyping(event.target as HTMLElement | null)
  const modal = isModalOpen()

  const candidates = registered.filter((s) => {
    if (typing && !s.keys.includes('mod+') && s.keys !== 'escape') return false
    if (modal && s.page) return false
    return true
  })

  const outcome = matchKeys(
    candidates.map((s) => s.keys),
    pendingChord,
    token,
  )

  if (chordTimer) {
    clearTimeout(chordTimer)
    chordTimer = null
  }
  pendingChord = null

  if (outcome.pending) {
    pendingChord = outcome.pending
    chordTimer = setTimeout(() => {
      pendingChord = null
      chordTimer = null
    }, CHORD_TIMEOUT_MS)
    event.preventDefault()
    return
  }

  if (!outcome.run) return

  // The last one registered wins, so a page can shadow a global key while it is open.
  const shortcut = [...candidates].reverse().find((s) => s.keys === outcome.run)
  if (!shortcut) return

  event.preventDefault()
  shortcut.run(event)
}

let installed = false

/** Attaches the one keydown listener. Safe to call more than once. */
export function useKeyboard(): void {
  useEffect(() => {
    if (installed) return
    installed = true
    window.addEventListener('keydown', onKeyDown)

    return () => {
      window.removeEventListener('keydown', onKeyDown)
      installed = false
    }
  }, [])
}

export const IS_MAC = typeof navigator !== 'undefined' && /Mac|iPhone|iPad/.test(navigator.platform)
